sql-plugin/src/main/spark320/scala/com/nvidia/spark/rapids/shims/HybridFileSourceScanExecMeta.scala

```diff
@@ -0,0 +1,216 @@
+                                   parent: Option[RapidsMeta[_, _, _]],
+                                   rule: DataFromReplacementRule)
+  extends SparkPlanMeta[FileSourceScanExec](plan, conf, parent, rule) {
+
+  // Replaces SubqueryBroadcastExec inside dynamic pruning filters with native counterpart
+  // if possible. Instead regarding filters as childExprs of current Meta, we create
+  // a new meta for SubqueryBroadcastExec. The reason is that the native replacement of
+  // FileSourceScan is independent from the replacement of the partitionFilters.
+  private lazy val partitionFilters = {
+    val convertBroadcast = (bc: SubqueryBroadcastExec) => {
+      val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
+      meta.tagForExplain()
+      meta.convertIfNeeded().asInstanceOf[BaseSubqueryExec]
+    }
+    wrapped.partitionFilters.map { filter =>
+      filter.transformDown {
+        case dpe@DynamicPruningExpression(inSub: InSubqueryExec) =>
+          inSub.plan match {
+            case bc: SubqueryBroadcastExec =>
+              dpe.copy(inSub.copy(plan = convertBroadcast(bc)))
+            case reuse@ReusedSubqueryExec(bc: SubqueryBroadcastExec) =>
+              dpe.copy(inSub.copy(plan = reuse.copy(convertBroadcast(bc))))
+            case _ =>
+              dpe
+          }
+      }
+    }
+  }
+
+  // partition filters and data filters are not run on the GPU
+  override val childExprs: Seq[ExprMeta[_]] = Seq.empty
+
+  override def tagPlanForGpu(): Unit = {
+    val cls = wrapped.relation.fileFormat.getClass
+    if (cls != classOf[ParquetFileFormat]) {
+      willNotWorkOnGpu(s"unsupported file format: ${cls.getCanonicalName}")
+    }
+  }
+
+  override def convertToGpu(): GpuExec = {
+    // Modifies the original plan to support DPP
+    val fixedExec = wrapped.copy(partitionFilters = partitionFilters)
+    HybridFileSourceScanExec(fixedExec)(conf)
+  }
+}
+
+object HybridFileSourceScanExecMeta {
+  private val HYBRID_JAR_PLUGIN_CLASS_NAME = "com.nvidia.spark.rapids.hybrid.HybridPluginWrapper"
+
+  // Determines whether using HybridScan or GpuScan
+  def useHybridScan(conf: RapidsConf, fsse: FileSourceScanExec): Boolean = {
+    val isEnabled = if (conf.useHybridParquetReader) {
+      require(conf.loadHybridBackend,
+        "Hybrid backend was NOT loaded during the launch of spark-rapids plugin")
+      true
+    } else {
+      false
+    }
+    // For the time being, only support reading Parquet
+    lazy val isParquet = fsse.relation.fileFormat.getClass == classOf[ParquetFileFormat]
+    // Check if data types of all fields are supported by HybridParquetReader
+    lazy val allSupportedTypes = fsse.requiredSchema.exists { field =>
+      TrampolineUtil.dataTypeExistsRecursively(field.dataType, {
+        // For the time being, the native backend may return incorrect results over nestedMap
+        case MapType(kt, vt, _) if kt.isInstanceOf[MapType] || vt.isInstanceOf[MapType] => false
```

Comment:
What about if it is a MapType, but kt or vt is not directly a map, but might be a LIST of MAP, so a struct with a MAP in it? Do we know the cause of this error so that we can limit things properly? If not then I would rather just stick with a MAP at the top level and any nested maps are not allowed.

Also what happens if the data is a LIST<STRUCT> Internally in Parquet a Map is just a LIST<STRUCT<KEY, VALUE>> would we have similar issues if we had one of them be nested?

Reply:
Hi @revans2 , I am sorry that I did not audit carefully on which types is unsupported  by native backend. Just before, I ran a rather comprehensive test:
```python
hybrid_gens_test = [
     # failed
    [decimal_gen_32bit_neg_scale],
    [decimal_gen_128bit],
    decimal_64_map_gens,
    [MapGen(TimestampGen(nullable=False), ArrayGen(string_gen))],
    [MapGen(RepeatSeqGen(IntegerGen(nullable=False), 10), TimestampGen())],
    [MapGen(RepeatSeqGen(IntegerGen(nullable=False), 10), decimal_gen_32bit)],
    [MapGen(RepeatSeqGen(IntegerGen(nullable=False), 10), decimal_gen_64bit)],
    # failed
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), decimal_gen_128bit)],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), ArrayGen(string_gen))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), ArrayGen(ArrayGen(long_gen)))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), ArrayGen(ArrayGen(string_gen)))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False),
            StructGen([['child0', string_gen],
                       ['child1', double_gen],
                       ['child2', int_gen],
                       ['child3', StructGen([['child0', ArrayGen(byte_gen)],
                                             ['child1', byte_gen],
                                             ['child2', float_gen],
                                             ['child3', decimal_gen_64bit]])]]))
     ],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False),
            StructGen([['child0', ArrayGen(ArrayGen(long_gen))],
                       ['child1', ArrayGen(string_gen)],
                       ['child2', ArrayGen(ArrayGen(string_gen))]]))
     ],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False),
            ArrayGen(MapGen(LongGen(nullable=False), long_gen)))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False),
            ArrayGen(MapGen(IntegerGen(nullable=False), string_gen)))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False),
            ArrayGen(ArrayGen(MapGen(IntegerGen(nullable=False), string_gen))))],

    [ArrayGen(ArrayGen(string_gen))],
    [ArrayGen(ArrayGen(long_gen))],
    # failed
    [ArrayGen(MapGen(LongGen(nullable=False), long_gen))],
    # failed
    [ArrayGen(MapGen(StringGen(pattern='key_[0-9]', nullable=False), long_gen))],

     # failed
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), MapGen(LongGen(nullable=False), long_gen))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), MapGen(LongGen(nullable=False), string_gen))],
    [MapGen(StringGen(pattern='key_[0-9]', nullable=False), simple_string_to_string_map_gen)],
     # failed
    [StructGen([['child0', MapGen(LongGen(nullable=False), long_gen)],
                ['child1', MapGen(StringGen(pattern='key_[0-9]', nullable=False), long_gen)],
                ['child2', MapGen(IntegerGen(nullable=False), decimal_gen_64bit)],
                ['child3', StructGen([["cc", MapGen(IntegerGen(nullable=False), decimal_gen_32bit)]])]
                ]),
     ],
    [StructGen([['cc', MapGen(IntegerGen(nullable=False), decimal_gen_64bit)]])],
     # failed
    [StructGen([['cc', ArrayGen(MapGen(IntegerGen(nullable=False), string_gen))]])],
    [StructGen([['cc', ArrayGen(ArrayGen(MapGen(IntegerGen(nullable=False), string_gen)))]])],
]```

The test result suggested the unsupported types are:
1. Decimal with negative scale is NOT supported
2. Decimal128 inside nested types is NOT supported
3. BinaryType is NOT supported
4. MapType inside nested types (Struct of Map/Array of Map/Map of Map) is NOT fully supported

I reworked the typeCheck function are integration tests according to the new finding.